### PR TITLE
fix next up link to work with nested articles

### PR DESCRIPTION
**Pull request by piyalbasu**
My original implementation didn't take nested articles into account. whoops!

### CODE REVIEW

File: src/helpers/documentation.js [hunk 1]
```diff
@@ -235,9 +209,85 @@ export const buildDocsContents = (data, rootDir) => {
     insertPageData(relPath, contents, articles, rootPageData);
   });
 
+  // Now that content is in proper order, travel through tree and add Next Up link to each article
```
Comment (piyalbasu) [inline src/helpers/documentation.js]: my mistake before was adding the next up links before we properly grouped all the data. That's why things were in the right order in the left nav, but the next up links for nested items did not work at all

File: src/helpers/documentation.js [hunk 2]
```diff
@@ -235,9 +209,85 @@ export const buildDocsContents = (data, rootDir) => {
...
 };
 
+/**
+ * addNextUpToArticles recursively travels through contents tree to add a Next Up property to each
+ * @param {array} articles List of articles of the current topic.
+ * @param {number} articleIndex The index of the article we're working with
+ * @param {number} nextTopic List of articles of the next topic
+ * @returns {void}
+ */
+const addNextUpToArticles = (articles, articleIndex, nextTopic) => {
```
Comment (piyalbasu) [inline src/helpers/documentation.js]: take each section of articles, and go through each one and add the next up link to each one.

if we encounter a nested group of articles, run through this func with that group

File: src/helpers/documentation.js [hunk 3]
```diff
@@ -235,9 +209,85 @@ export const buildDocsContents = (data, rootDir) => {
...
+};
+
+/**
+ * nextUpLink returns the url and title of the next article, whether it's in the same topic or if it's in the next topic
+ * @param {array} topicArticles List of articles of the current topic.
+ * @param {number} articleIndex The index of the article we're working with
+ * @param {number} nextTopic List of articles of the next topic
+ * @returns {object} { title, url }
+ */
+export const nextUpLink = (topicArticles, articleIndex, nextTopic) => {
```
Comment (piyalbasu) [inline src/helpers/documentation.js]: figure out if the next article is a sibling or if the next article is the first child of the next topic

File: src/helpers/documentation.js [hunk 4]
```diff
@@ -235,9 +209,85 @@ export const buildDocsContents = (data, rootDir) => {
...
+    url: nextNestedFirstArticle.url,
+  };
+};
+
+/**
+ * findNextFirstArticle recursively drills down nested articles to find the first article we can link to
+ * @param {object} articleData List of articles of the current topic.
+ * @returns {object} { title, url }
+ */
+const findNextFirstArticle = (articleData) => {
```
Comment (piyalbasu) [inline src/helpers/documentation.js]: this helper method just find the next article we can link to. in some cases, it's just the next sibling. but in others, it's a nested collection and we have to drill down to find an article

**Comment by jeesunikim** [comment]
Awesome! Thanks @piyalbasu

**Comment by vcarl** [comment]
Now that we have Jest set up, it'd be dope to get some unit tests in place for this @piyalbasu. Even if it's just a matter of extracting some fixture data and ensuring it transforms correctly, that'd be a nice boost to confidence if we have to touch this in the future.

**Pull request merged**